sdk/dotnet/Pulumi/Testing/MockMonitor.cs

```diff
@@ -0,0 +1,123 @@
+        {
+            var (id, state) = await _mocks.NewResourceAsync(request.Type, request.Name, ToDictionary(request.Object),
+                request.Provider, request.ImportId);
+            this.Resources.Add(resource);
+            return new RegisterResourceResponse
+            {
+                Id = id ?? request.ImportId,
+                Urn = NewUrn(request.Parent, request.Type, request.Name),
+                Object = await SerializeAsync(state)
+            };
+        }
+
+        public Task RegisterResourceOutputsAsync(RegisterResourceOutputsRequest request)
+        {
+            var stackUrn = NewUrn("", Stack._rootPulumiStackTypeName, $"{Deployment.Instance.ProjectName}-{Deployment.Instance.StackName}");
+            if (request.Urn == stackUrn)
+            {
+                StackOutputs = ToDictionary(request.Outputs);
+            }
+
+            return Task.CompletedTask;
+        }
+
+        public Task LogAsync(LogRequest request)
+        {
+            if (request.Severity == LogSeverity.Error)
+            {
+                this.Errors.Add(request.Message);
+            }
+            
+            return Task.CompletedTask;
+        }
+
+        public Task<SetRootResourceResponse> SetRootResourceAsync(SetRootResourceRequest request)
+        {
+            _rootResourceUrn = request.Urn;
+            return Task.FromResult(new SetRootResourceResponse());
+        }
+
+        public Task<GetRootResourceResponse> GetRootResourceAsync(
+            GetRootResourceRequest request)
+            => Task.FromResult(new GetRootResourceResponse { Urn = _rootResourceUrn });
+        
+        private static string NewUrn(string parent, string type, string name)
+        {
+            if (!string.IsNullOrEmpty(parent)) 
+            {
+                var qualifiedType = parent.Split("::")[2];
+                var parentType = qualifiedType.Split("$").First();
+                type = parentType + "$" + type;
+            }
+            return "urn:pulumi:" + string.Join("::", new[] { Deployment.Instance.StackName, Deployment.Instance.ProjectName, type, name });
+        }
+
+        private static ImmutableDictionary<string, object> ToDictionary(Struct s)
+        {
+            var builder = ImmutableDictionary.CreateBuilder<string, object>();
+            foreach (var (key, value) in s.Fields)
+            {
+                var data = Deserializer.Deserialize(value);
+                if (data.IsKnown && data.Value != null)
+                {
+                    builder.Add(key, data.Value);
+                }
+            }
+            return builder.ToImmutable();
+        }
+
+        private async Task<Struct> SerializeAsync(object o)
+        {
+            var dict = (o as IDictionary<string, object>)?.ToImmutableDictionary()
+                       ?? await _serializer.SerializeAsync("", o) as ImmutableDictionary<string, object>
+                       ?? ImmutableDictionary<string, object>.Empty;
```

Comment:
can we be more precise here, and throw if we get a value that we can't deal with?

Reply:
changed